http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/Http1TunnelConnectionPool.java

```diff
@@ -0,0 +1,165 @@
+        return acquire(eventLoop.newPromise());
+    }
+
+    @Override
+    public Future<Channel> acquire(Promise<Channel> promise) {
+        delegate.acquire(eventLoop.newPromise()).addListener((Future<Channel> f) -> {
+            if (f.isSuccess()) {
+                setupChannel(f.getNow(), promise);
+            } else {
+                promise.setFailure(f.cause());
+            }
+        });
+        return promise;
+    }
+
+    @Override
+    public Future<Void> release(Channel channel) {
+        return release(channel, eventLoop.newPromise());
+    }
+
+    @Override
+    public Future<Void> release(Channel channel, Promise<Void> promise) {
+        return delegate.release(channel, promise);
+    }
+
+    @Override
+    public void close() {
+        delegate.close();
+    }
+
+    private void setupChannel(Channel ch, Promise<Channel> acquirePromise) {
+        if (isTunnelEstablished(ch)) {
+            log.debug(() -> String.format("Tunnel already established for %s", ch.id().asShortText()));
+            acquirePromise.setSuccess(ch);
+            return;
+        }
+
+        log.debug(() -> String.format("Tunnel not yet established for channel %s. Establishing tunnel now.",
+                ch.id().asShortText()));
+
+        Promise<Channel> tunnelEstablishedPromise = eventLoop.newPromise();
+
+        SslHandler sslHandler = createSslHandlerIfNeeded(ch.alloc());
+        if (sslHandler != null) {
+            ch.pipeline().addLast(sslHandler);
+        }
+        ch.pipeline().addLast(initHandlerSupplier.newInitHandler(delegate, remoteAddress, tunnelEstablishedPromise));
+
+        tunnelEstablishedPromise.addListener((Future<Channel> f) -> {
+            if (f.isSuccess()) {
+                Channel tunnel = f.getNow();
+                handler.channelCreated(tunnel);
+                tunnel.attr(TUNNEL_ESTABLISHED_KEY).set(true);
+                acquirePromise.setSuccess(tunnel);
+            } else {
+                ch.close();
+                delegate.release(ch);
+
+                Throwable cause = f.cause();
+                log.error(() -> String.format("Unable to establish tunnel for channel %s", ch.id().asShortText()), cause);
+                acquirePromise.setFailure(cause);
+            }
+        });
+    }
+
+    private SslHandler createSslHandlerIfNeeded(ByteBufAllocator alloc) {
+        if (sslContext == null) {
+            return null;
+        }
+
+        String scheme = proxyAddress.getScheme();
+
+        if (!"https".equals(StringUtils.lowerCase(scheme))) {
+            return null;
+        }
+
+        return sslContext.newHandler(alloc, proxyAddress.getHost(), proxyAddress.getPort());
+    }
+
+    private static boolean isTunnelEstablished(Channel ch) {
+        Boolean established = ch.attr(TUNNEL_ESTABLISHED_KEY).get();
+        return Boolean.TRUE.equals(established);
+    }
+
+    @SdkTestInternalApi
+    @FunctionalInterface
+    interface InitHandlerSupplier {
```

Comment:
Minor: any reason this needs to be an inner interface?

Reply:
Not really, this was mainly to enable easier testing of the pool.